File: 2.两数相加.js
/*
 * @lc app=leetcode.cn id=2 lang=javascript
 *
 * [2] 两数相加
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    let nln = new ListNode();
    //使用链表的引用来对链表操作，如果直接对链表操作，会摧毁链表的结构。
    let middle = nln;
    let tenIf = 0;
    while(l1 !== null || l2 !== null) {
        let sum =  0;
        if(l1 !== null){
            sum += l1.val;
            l1 = l1.next;
        }
        if(l2 !== null){
            sum += l2.val;
            l2 = l2.next;
        }
        sum += tenIf;
        middle.next = new ListNode(sum % 10);
        tenIf = Math.floor(sum / 10);
        middle = middle.next;
    }

    if(tenIf > 0){
        middle.next = new ListNode(tenIf);
        //在为了更加方便地对链表写入数据，使用了空头链表  返回时，需将空头链表给移除！
        return nln.next;
    }
    return nln.next;
};
// @lc code=end

File: c/2.两数相加.c
/*
 * @lc app=leetcode.cn id=2 lang=c
 *
 * [2] 两数相加
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2){
    //储存进位
    int carry = 0;
    // 创造一个头结点，让头指针dummy指向它
    struct ListNode* dummy = malloc(sizeof(struct ListNode));
    // 创造一个指针，让cur指向最后
    struct ListNode* cur = dummy;

    //他们三个中有一个还有值 进行下一轮循环
    while(l1 || l2 || carry) {
        if(l1) {
            carry += l1->val;
            l1 = l1->next;
        }

        if(l2) {
            carry += l2->val;
            l2 = l2->next;
        }

        cur->next = malloc(sizeof(struct ListNode));
        cur->val = carry % 10;
        cur->next = (void *)0;

        carry /= 10;
    }

    return dummy->next;
}
// @lc code=end

